### ISSUE TITLE
DCR: State Management API Overhaul

### ISSUE THREAD

**Issue by drub0y** [Microsoft/botbuilder-dotnet#375]
# TL;DR;

Refactor the state management APIs to address discoverability issues and light up scenarios that are not possible with the existing design.

## Overview

We have received consistent feedback from partners about the discoverability/friction of setting up and working with the current state management APIs. My perspective comes from the .NET side since I'm most familiar with the typical patterns people expect there and I can understand why there is some difficultly in understanding the existing design, but I have heard of some confusion on the JS side as well.

Additionally there are some scenarios that have been raised that are simply not possible to achieve with the APIs today either due to bugs or design limitations.

Some of the feedback we've received includes:

* Inability to control multiple pieces of conversation/user state
* Have to register multiple pieces of middleware to work with each different type of state
* Coupling of core state management responsibilities with middleware class feels off
* Inability to support multiple pieces of state of same type under different keys in the same state space
* `FlexObject`/`StoreItem`/`StoreItems` APIs are confusing and certain API usages currently strip the strong types off due to use of `dynamic`
* Use of static methods on ancillary state classes (e.g. `ConversationState<SomeState>.Get(context)`) inhibit discovery and are not a familiar .NET pattern

## Goals

The goals of this DCR are to address the aforementioned issues as follows:

* Do not constrain how many types of state can be stored:
* Support _N_ state "namespaces"
  * Conversation
  * User
  * Bot
  * Custom
* Each namespace supports _M_ key/values
* Support multiple instances of the same state value type with differing keys
* Middleware should be able to store its own state, completely isolated from the bot's state
* Options for optimized loading/saving of state:
  * All state for a namespace
  * All state for all namespaces
  * Specific subset of state
* Remove use of `dynamic` altogether; state should be pure POCOs and not require subtyping any specific interface/class
* Auto-loading/saving state
  * All keys for all namespaces
  * All keys for a subset namespaces
  * A subset of keys from a subset of namespaces
  * Auto loading/saving would be concurrent per _N_ state namespaces
~* Normalize the handling of optimistic concurrency violations so that those interacting with the state management APIs can be agnostic of the underlying store~ -- optimistic concurrency support removed [per discussion below](#OptimisticConcurrencyOpenIssue)

## Proposal

> **NOTE:** All naming proposed here within is open for debate.

I am proposing we support an infinite number of "namespaces" for state that allow an infinite number of key/values per namespace. At its simplest, you can imagine this as a dictionary of dictionaries. The first dictionary is keyed by a given state namespace and then the nested dictionary for that state namespace is keyed by whatever key scheme that namespace has chosen to impose.

The default, well-known namespaces that will come "in the box" would be:

| Type  | Proposed Namespace |
|---|---|
| Conversation State  | `/channels/{channelId}/conversations/{conversationId}`  |
| User State  |  `/channels/{channelId}/users/{userId}` |
| Conversation Specific User State  | `/channels/{channelId}/conversations/{conversationId}/users/{userId}`
| Bot State | `/bot/{botId}`  |

### Open Questions around namespaces

* Are there any others that we should bake in?
* Should we impose a URN scheme for namespaces?
* Should we prefix all of the built-in namespaces with something like `bot-framework:`?

### State Stores

> **NOTE:** Most users will not work directly with this abstraction, it exists as the backing storage mechanism for state managers which will be dicussed next.

The existing `BotState<StateT>` assumes a `dynamic` bag of state in the form of `StoreItems` a given set of state. There is currently concept of multiple key/values for a given store entry. Support for `ETag` is also something that the `StateT` must itself directly implement by subtyping `IStoreItem` which burdens the developer and pollutes their POCOs unecessarily. Finally, `BotState<StateT>` is a piece of middleware in addition to its core storage responsibilities which violates single responsibility principle.

To address these issues, the new approach starts with an `IStateStorageProvier` interface that replaces `IStore`. An `IStateStorageEntry` interface is also introduced to hide the underlying storage represenation and details such as `ETag`. Finally, `IStateStorageProvider` is not middleware, instead [there is a separate `StateManagementMiddleware`](#StateManagementMiddleware).

```CSharp
public interface IStateStorageProvider
{
    IStateStorageEntry CreateNewEntry(string stateNamespace, string key);

    Task<IEnumerable<IStateStorageEntry>> Load(string stateNamespace);

    Task<IStateStorageEntry> Load(string stateNamespace, string key);

    Task<IEnumerable<IStateStorageEntry>> Load(string stateNamespace, IEnumerable<string> keys);

    Task Save(IEnumerable<IStateStorageEntry> stateStorageEntries);

    Task Delete(string stateNamespace);

    Task Delete(string stateNamespace, IEnumerable<string> keys);
}

public interface IStateStorageEntry
{
    string Namespace { get; }

    string Key { get;  }

    T GetValue<T>() where T : class, new();

    void SetValue<T>(T value) where T : class;
}
```

~~>**NOTE**: I am strongly considering renaming this interface to `IStateStorageProvider` based on other naming patterns in the .NET ecosystem.~~ -- _Changed_

#### `IStateStorageProvider` implementations

At minimum we would provide the following state store implementations "in the box":

| Implementation | Description |
|---|---|
|`MemoryStateStorageProvider` | An in memory implementation that is thread safe. Literally employs  the dictionary of dictionaries approach. Great for local dev and stage demos, _not_ realistic for production. |
|`FileSystemStateStorageProvider` | A file system based implementation that, given a base path, stores the data for each state namespace under its own subfolder and each key/value in an individual file. It is currently hardcoded to store using `JSON` UTF-8 text, but the serialization format could theoretically be abstracted. This is also good for local dev and stage demos where state persistence across runs is desired. It could also be a viable solution for production systems, though there are likely more scalable and performant solutions available. |
|`AzureTableStorageStateStorageProvider` | An Azure Table Storage based implementation which, _surprise_ 🎊, stores state in a given Azure Table Storage table. It leverages the state namespace as the `PartitionKey` and the key as the `RowKey`. It uses the built in "flattening" capabilities of the Azure Table Storage SDK to flatten state POCOs out into individual columns. You can use it for local dev and stage demos with the Azure Storage SDK emulator. Of these three implementations, it's the most appropriate option for a production system.  |

##### Deferred state materialization

Materializing a `TState` class from an underlying data store can be costly and introduce a lot of garbage to be collected. When you consider bulk load scenarios, it doesn't always make sense to pay the price of materialization up front if only some small subset of the bulk loaded state is actually used. To help stores with this there is an additional class called `DeferredStateStorageEntry` which allows store implementations to defer the materialization of the underlying representation (e.g. byte array, memorystream, table entities, etc.) into a `TState` to only happen when that `TState` is actually first requested by a `StateManager`.

~##### Optimistic Concurrency

The proposed design enables optimistic concurrency at its core by exposing the concept of an `ETag` on `StateStorageEntry` which `IStateStorageProvider` implementations can use to determine if they are dealing with the latest version of the data and provide optimistic concurrency guarantees when attempting to apply updates to the underlying data store.

Also included is a common `StateOptimisticConcurrencyViolationException` class which is expected to be thrown when a particular implementation runs into a violation.~ -- feature removed [per explanation below discussion below](#OptimisticConcurrencyOpenIssue)

#### State Storage Open Issues

<a name="OptimisticConcurrencyOpenIssue"></a>~* Should optimistic concurrency be _required_? Right now the concept of `ETag` is present on the `IStateStorageEntry` interface and implies somehow that all stores _might_ provide optimistic concurrency. This is by no means easy (e.g. file system), nor is it always necessary for people, sometimes Last Write Wins is just fine.~
  * ~Even for stores that support it, we should allow people to opt-out of it and just use a Last Write Wins model and we want that configuration to be consistent for all `IStateStorageProvider` implementations.~ -- After some internal debate, we are removing support for optimistic concurrency (i.e. `ETag`) from the state management APIs. It was deemed unnecessary due to the existing partitioning that naturally takes place in a conversation which should avoid most conflicts and because there's no good solution for what to do with the fact that an optimistic concurrency violation occurs. Instead favoring what would normally happen in a conversation which would be opportunities for confirmation and clarification with the user who would then have a chance to notice something might be out of alignment and correct it.

##### Read Through State Storage Scenario

Since `IStateStorageProvider` is an abstraction, one could imagine a `ReadThroughStateStorageProvider` implementation which composes two different state storage implementations together and reads only from the first store and reads through to the second store if the value is not provided by the first store. More specifcally imagine a `RedisStateStorageProvider` being configured "in front of" an `AzureTableStorageStateStorageProvider`. When a write happens it writes through both stores. When a read happens it first tries to read from the `RedisStateStorageProvider` and if that returns the requested entry then that's it, no hit will be made to the underlying `AzureTableStorageStateStorageProvider`. If the value is not found in the `RedisStateStorageProvider`, then it proceeds to reads through to the underlying `AzureTableStorageStateStorageProvider` and, assuming if it finds the value there, the value will be placed into the `RedisStateStorageProvider` for subsequent reads to find.

### State Managers

> *NOTE:* This is the API that middleware and bots will interact with directly to perform CRUD operations on their state.

Today there is no direct mapping for this concept. Today you would simply get your state object back by calling the appropriate static class method (e.g. `ConversationState<MyState>.Get(myContext)`). Aside from consistent reports that this isn't very discoverable, it's also limited to only a single `MyState` and there is also a bug where you really can't have more than one `CoversationState<StateT>` registered today or the last one effectivly "wins".

The proposal is to introduce an `IStateManager` interface that acts as the abstraction for managing state for a given namespace. It offers direct control over loading state and saving state changes.

```CSharp
public interface IStateManager
{
    Task<TState> Get<TState>(string key) where TState : class, new();

    void Set<TState>(string key, TState state) where TState : class;

    void Delete(string key);

    Task LoadAll();

    Task Load(IEnumerable<string> keys);

    Task SaveChanges();
}
```

The interface actually exists mainly to enable mocking scenarios because a concrete implementation of the interface is provided in the box via the `StateManager` class. No one is expected to implement their own `IStateManager` today. In fact, right now, there is no way to override the built in `StateManager` implementation, but it would be entirely possible to enable this in the future by introducing a provider model.

In addition to `IStateManager` there are more interfaces representing common state scenarios even more strongly:

```CSharp
public interface IConversationStateManager : IStateManager
{
    string ChannelId { get; }
    string ConversationId { get; }
}

public interface IUserStateManager : IStateManager
{
    string UserId { get; }
}
```

Likewise, there are concrete classes for these in the form of `ConversationStateManager` and `UserManager`.

> **NOTE:**  There is currently no `[I]BotStateManager` yet, but I do intend to add one.

#### StateManager -> StateStorageProvider relationship

A `StateManager` instance is bound to a single `IStateStorageProvider`. Multiple `StateManager`s can be bound to the same store. Multiple `StateManager`s can be bound to multiple `IStateStorageProvider`s in the sample app. The most common scenario is a single state store with mutliple state managers which looks something like this:

<pre>
 +----------------------------+ +---------------------+
 |                            | |                     |
 |  ConversationStateManager  | |  UserStateManager   |
 |                            | |                     |
 +------+-------------^-------+ +-----+-----------^---+
        |             |               |           |
        |             |               |           |
        |  namespace  |               |           |
/channels/123/conversations/abc       |           |
        |             |               |           |
        |             |               | namespace |
        |             |              /users/jdoe123
        |             |               |           |
        |             |               |           |
 +------v-------------+---------------v-----------+---+
 |                                                    |
 |        AzureTableStorageStateStorageProvider       |
 |                                                    |
 +----------------------------------------------------+
</pre>

Other more complex scenarios might include:

##### Scenario A -  Multiple Stores

I want to store my conversation state in Azure Table Storage and my user state in Azure Active Directory.

<pre>
      +----------------------------+                  +---------------------+
      |                            |                  |                     |
      |  ConversationStateManager  |                  |  UserStateManager   |
      |                            |                  |                     |
      +------+-------------^-------+                  +-----+-----------^---+
             |             |                                |           |
             |             |                                |           |
             | namespace   |                                |           |
     /channels/123/conversations/abc                        |           |
             |             |                                |           |
             |             |                                | namespace |
             |             |                               /users/jdoe123
             |             |                                |           |
             |             |                                |           |
 +-----------v-------------+-------------+  +---------------v-----------+---------------+
 |                                       |  |                                           |
 | AzureTableStorageStateStorageProvider |  |  AzureActiveDirectoryStateStorageProvider |
 |                                       |  |                                           |
 +---------------------------------------+  +-------------------------------------------+

</pre>

##### Scenario B -  Multiple Stores, some shared

I want to store my conversation state and user state in Redis, but my bot state in Azure Table Storage.

<pre>
 +----------------------------+ +---------------------+     +---------------------+
 |                            | |                     |     |                     |
 |  ConversationStateManager  | |  UserStateManager   |     |   BotStateManager   |
 |                            | |                     |     |                     |
 +------+-------------^-------+ +-----+-----------^---+     +-----+-----------^---+
        |             |               |           |               |           |
        |             |               |           |               |           |
        |  namespace  |               |           |               |           |
/channels/123/conversations/abc       |           |               |           |
        |             |               | namespace |               |           |
        |             |              /users/jdoe123               |           |
        |             |               |           |               | namespace |
        |             |               |           |               /bot/mybot131
        |             |               |           |               |           |
 +------v-------------+---------------v-----------+---+ +---------v-----------+-------+
 |                                                    | |                             |
 |         AzureTableStorageStateStorageProvider      | |  RedisStateStorageProvider  |
 |                                                    | |                             |
 +----------------------------------------------------+ +-----------------------------+
</pre>

#### State caching

A `StateManager` acts as the read-through cache for a given namespace in the underlying `IStateStorageProvider` for the lifetime of the turn. Multiple calls to the same instance of a `IStateManager`'s `Get<TState>(string key)` _MUST_ always yield the same `TState` instance for the lifetime of the turn. Once the turn is over, any state changes that are intended to be kept should be persisted with a call to `IStateManager::SaveChanges()` so that when the next turn happens, which could be in a completely different process or on another node entirely, the state can be rehydrated from the underlying storage provider.

```mermaid
sequenceDiagram
    participant B as Bot
    participant SM as State Manager
    participant IC as Internal Cache
    participant SS as State Storage Provider
    B->>SM: Get&lt;SomeState&gt;()
    SM->>IC: Check cache
    alt is already cached
        IC-->>SM: Found cached state
        SM-->>B: Return state
    else not yet cached
        SM->>SS: Load from underlying store
        SS-->>SM: Requested state entry/entries
        SM->>IC: Place in cache
        SM-->>B:Return state
    end
```

>**NOTE:** the above sequence diagram is in [Mermaid](https://github.com/knsv/mermaid/) format. [There is a plugin that renders this kind of Markdown in VS Code](https://marketplace.visualstudio.com/items?itemName=bierner.markdown-mermaid) or you can [go view it here on the web](https://mermaidjs.github.io/mermaid-live-editor/) (use Chrome, doesn't work in Edge 😢).

#### State loading

Individual pieces of state are retrieved through calls to `IStateManager::Get<TState>(string key)` on a key by key basis. By default, if the state is not already loaded into the `StateManager`, it will go call `IStateStorageProvider::Load(string stateNamespace, string key)` which might translate to an I/O operation for that single piece of state.

Bulk loading APIs are also offered via the `LoadAll` and `Load(IEnumerable<string> keys)` methods which enable the developer to optimize scenarios by giving the state manager a chance to minimize the number of interactions with the underlying storage implementation. 

For example, a call to `IStateManager::LoadAll` on a state manager backed by an `AzureTableStorageStateStorageProvider` this can mean making a single query to load all rows in a given partition and stream them back in one shot rather than making _N_ queries over time.

#### State change tracking

As of this version of the API there is no automatic change tracking. If you want a particular piece of state to be persisted, you _MUST explicitly_ tell the state manager via  `IStateManager::Set<TState>(TState state)`.

##### Open discussion

* I don't want to have to call `Set`... why is this hard? 😒

#### Saving Changes

There is a single `IStateManager::SaveChanges()` which always saves all pending changes in the instance which are known to be: new, updated or deleted values.

#### Sample Usage of "raw" StateManager APIs

The following samples show how one would work with the "raw" StateManager APIs. We would not expect developers to write code like this themselves the majority of the time; instead they would use the `ITurnContext` extensions ([see below for more on this](#ITurnContextExtensions)).

##### Create a StateManager for a custom namespace backed by a MemoryStateStorageProvider

```CSharp
var stateManager = new StateManager("MyCustomStateNamespace", new MemoryStateStorageProvider());
```

##### Create a ConversationStateManager backed by a FileSystemStateStorageProvider and get a piece of state from it

```CSharp
var stateStorageProvider = new FileSystemStateStorageProvider(@"\\someserver\MyBotState"));
var stateManager = new ConversationStateManager(
    turnContext.Activity.ChannelId,
    turnContext.Activity.Conversation.Id,
    stateStorageProvider);

var state = await stateManager.Get<SomeState>();
```

##### Create a UserStateManager backed by an AzureTableStorageStateStorageProvider, add a new piece of state to it and save the changes

```CSharp
var azureCloudStorageAccount = GetCloudStorageAccount();
var stateStorageProvider = new AzureTableStorageStateStorageProvider(
    azureCloudStorageAccount,
    "MyUserStateTableName");
var stateManager = new UserStateManager(
    turnContext.Activity.From.Id,
    stateStorageProvider);

stateManager.Set(
    "MyStateKey", 
    new SomeState
    {
        MyNumericProperty = 1,
        MyStringProperty = "Hello World!",
    });

await stateManager.SaveChanges();
```

### <a name="ITurnContextExtensions">ITurnContext extensions</a>

The state management APIs introduce a set of extension methods for `ITurnContext` which are the top level APIs that middleware and bot logic will call to get themselves an `IStateManager` for a given namespace.

In the box we provide three extensions:

| Extension Method | Explanation |
|---|---|
| `ConversationState()` | returns the well-known `IConversationStateManager` implementation |
| `UserState()` | returns the well-known `IUserStateManager` implementation |
| `BotState()` | returns the well-known `IBotStateManager` implementation (_NOTE: to be added_) |
| `State(string namespace)` | returns an `IStateManager` that represents the specified state namespace |

#### Sample ITurnContext extensions usage

This set of samples now uses the friendly extension method APIs that help abstract you from the lower level details in the previous examples.

##### Get a piece of conversation state

```CSharp
var state = await turnContext.ConversationState().Get<SomeMiddlewareState>();
```

##### Get a named piece of user state

```CSharp
var state = await turnContext.UserState().Get<SomeUserState>("SomeUserState-1");
```

##### Get a piece of state from a custom namespace

```CSharp
var stateManager = turnContext.State("AReallyAwesomeProprietaryStateNamespace");
var state = await stateManager.Get<AwesomeState>();

state.NumberOfMessages++;
stateManager.Set(state);
await stateManager.SaveChanges();
```

##### Get and update a piece of conversation state

```CSharp
var conversationStateManager = turnContext.ConversationState();
var state = await conversationStateManager.Get<SomeState>();

state.NumberOfMessages++;
conversationStateManager.Set(state);
```

##### Get a piece of conversation state, update it and explicitly save the changes

```CSharp
var conversationStateManager = turnContext.ConversationState();
var state = await conversationStateManager.Get<SomeState>();

state.NumberOfMessages++;
conversationStateManager.Set(state);

await conversationStateManager.SaveChanges();
```

#### Possible API changes

There has been some internal discussion about consolidating to a single extension method to reduce "noise" added to the context and then that would return an interface which then exposes all things related to state from there. That might look something like this instead:

```CSharp
public interface IStateManagementTurnService
{
    IConversationStateManager ConversationState { get; }
    IUserStateManager UserState { get; }
    IBotStateManager BotState { get; }
    IStateManager ForNamespace(string namespace); 
}

public static class StateManagementTurnContextExtensions
{
    public static IStateManagementTurnService State(this ITurnContext turnContext) => turnContext.Services.Get<IStateManagementTurnService>();
}
```

And then, using one of samples from above, the API would look like this instead:

###### Get a piece of conversation state (revised)

```CSharp
var state = await turnContext.State().ConversationState.Get<SomeState>();
```

###### Get a piece of state from a custom namespace (revised)

```CSharp
var state = await turnContext.State().ForNamespace("AReallyAwesomeProprietaryStateNamespace").Get<SomeState>();
```

### <a name="StateManagementMiddleware">State Management Middelware</a>

With the current API design you need to configure multiple pieces of middleware in the form of `BotState<StateT>` instances, one for each type of state that you wish to store in each particular implicit namespace. There is also no configuration of auto-loading/saving; it just happens whether you want it to or not and with no control over any granularity (i.e. only load some subset of state). Finally, because each type of state is configured as a separate piece of middleware, there is no opportunity for concurrent loading/saving due to the step-based nature of middleware execution.

The proposed API offers a single piece of middleware called the `StateManagementMiddleware`. A developer configures all aspects of state management via this one middleware (e.g. stores, managers, auto-loading/saving).

> **NOTE:** I am still fleshing out the configuration aspects of this middleware.

#### State Store Configuration

A default `IStateStorageProvider` can be configured as well as _N_ named `IStateStorageProvider` instances:

###### Configure a `MemoryStateStorageProvider` as the default store

```CSharp
middleware.Use(
    new StateManagementMiddleware()
        .UseDefaultStore(new MemoryStateStorageProvider()));
```

###### Configure multiple named stores

```CSharp
middleware.Use(
    new StateManagementMiddleware()
        .UseStore("MyFirstStore", new MemoryStateStorageProvider())
        .UseStore("MySecondStore", new AzureTableStorageStateStorageProvider(/* ... */)));
```

#### State Manager Configuration

You can also configure _N_ `IStateManagers`. Each state manager has its own individual configuration options. By default you don't need to configure anything and it will just link the state manager up to whatever state store is configured by default:

##### Configure conversation state w/no explicit options

```CSharp
middleware.Use(
    new StateManagementMiddleware()
        .UseDefaultStore(new MemoryStateStorageProvider());
        .UseConversationState());
```

##### Configure conversation state w/no explicit options and user state with a specific, named store

```CSharp
middleware.Use(
    new StateManagementMiddleware()
        .UseDefaultStore(new MemoryStateStorageProvider());
        .UseStore("MyUserStateStorageProvider", new AzureTableStorageStateStorageProvider(/* ... */))
        .UseConversationState()
        .UseUserState(config =>
        {
            config.UseStore("MyUserStateStorageProvider");
        }));
```

##### Configure a custom state namespace

```CSharp
middleware.Use(
    new StateManagementMiddleware()
        .UseDefaultStore(new MemoryStateStorageProvider());
        .UseState("MyCustomNamespace")));
```

#### Auto-load/save Configuration

You can configure automatic loading/saving of all or some explicitly configured subset of state namespaces and all or some configured subset of keys within those namespaces. 

> **NOTE:** For performance, the middleware will do this for each state manager in parallel.

##### Auto-load/save all state from all configured state managers

```CSharp
middleware.Use(
    new StateManagementMiddleware()
        // ... state store and manager configuration here ...
        .AutoLoadAll()
        .AutoSaveAll());
```

##### Auto-load only conversation state

```CSharp
middleware.Use(
    new StateManagementMiddleware()
        .UseConversationState(config =>
        {
            config.AutoLoadAll();
        }));
```

##### Auto-load a particular custom namespace only

```CSharp
middleware.Use(
    new StateManagementMiddleware()
        .UseState(
            "SomeMiddlewareStateNamespace",
            config =>
            {
                config.AutoLoadAll();
            })
);
```

##### Auto-load all conversation state and only a certain subset of some custom namespace

```CSharp
middleware.Use(
    new StateManagementMiddleware()
        .UseConversationState(config =>
        {
            config.AutoLoadAll();
        })
        .UseState(
            "MyCustomStateNamespace",
            config =>
            {
                config.AutoLoad(
                    "StateKeyA",
                    "StateKeyF",
                    // ... etc ...
                );
            })
);
```

#### Open Issues

* Auto-saving is currently all or nothing. I can't personally think of a framework where I've used or even seen it possible to partially auto-save some subset of state. Usually you want it all saved automatically or you don't. Happy to be convinced otherwise and isn't _hard_ to add at all... just not sure it's worth it.
* Not sure how I feel about having to configure each well-known state (e.g. conversation, user, bot) explicitly. Might make sense to have a convenience method that just registers them all for the masses, but then again it might hide too much and also would inflict a performance penalty if not using a particular type of state and you also configured auto-load/save.

**Comment by garypretty** [comment]
Thanks @drub0y. This all aligns really well with the conversations we have been having and the proposal I was putting together. It really is awesome stuff.  A few comments from me;

* Having a separate StateManagementMiddleware :+1 Fantastic. 

* Regarding the open question around state change tracking and explicitly calling Set ("why is this so hard?") - I echo your sentiments here.  I would be interested to know if this was implemented due to specific feedback, but right now if I am calling Get then I would expect to call Set. I favor the explicit approach.

* The default state store is a nice idea, but I wonder if this will lead to people forgetting to add the default store and then not explicitly referencing one when adding middleware.  I kind of favor the traditional approach of specifying your state store for each piece of state middleware if I am honest. At least this way the user cannot fall into the trap of not specifying one and means its also really clear which store is being used for each piece of middleware.

* Totally fine with auto-save being all or nothing.  This makes sense to me and, likewise, I cannot see a situation where a more granular approach would be needed.

* As for the need to configure the well known states individually, I think this is ok.  There are many examples I can think of where you will need conversation state, but not user state and vice versa.

**Comment by drub0y** [comment]
> Regarding the open question around state change tracking and explicitly calling Set ("why is this so hard?") - I echo your sentiments here. I would be interested to know if this was implemented due to specific feedback, but right now if I am calling Get then I would expect to call Set. I favor the explicit approach.

Let's first make sure we can agree that we need a `Set` for _at least_ adding new instances of state to a namespace, right? In fact, I would change the method to `Add` explicitly if we do implement change tracking. Likewise, we need an explicit `Delete` because the intent should be clear that you're deleting. Sometimes you see some APIs allow passing a `null` value to a `Set` method to accomplish this, but I personally think that's a design smell.

So we're totally just talking about tracking changes for the purposes of updates and, just to level set, let's define what that change tracking is in this context: change tracking is all about knowing which pieces of state changed and then only issuing commands back to the underlying store to update those entities that have changed. The name of the game is efficiency. For example, if you pulled out `27` different pieces of state and you only changed `1` property on `1` of those pieces of state, you obviously don't want to be doing `26` writes that are effectively no-ops. Change tracking solves that.

**Disclaimer:** from here on I'm going to approach this as if people have no background in change tracking since I'm sure not everyone does.

Ok, so... what's the problem? Well, no approach to implicitly tracking changes on POCOs is "free". In fact, done wrong, it can be quite expensive. If you've ever followed the history of Entity Framework this has always been a huge deal for them for their model classes. They used to require you to inherit from a base `EntityObject` and trigger changes to underlying values through that. That made creating models a chore and better left to code-generation. Demand for supporting POCOs grew, so they eventually made the shift, but now they were left having to do snapshotting.

Snapshotting means you maintain some original representation of the state (perhaps just raw name/values), then when `SaveChanges` is called you compare the current version of each property of each piece of state against the original representation, figure out which once have a change and that's your delta for persistence. This works, but it comes with a cost: comparing `N` properties on `M` pieces of state. Now, one shortcut here is that if you can at least track which state objects have changed (i.e. a dirty bit) you can _at least_ avoid walking through all `M` of them and only work through that subset of `D⊆M` comparing the `N` properties on those.

For best performance EF has recently adopted a proxying technique. It doesn't require you inherit your model POCOs from any base class or implement any interface, but it does require you to leave them unsealed and make all the properties `virtual`. Then, at runtime, it will subclass your POCO, override all the properties to do the change tracking right on that instance. Not only that, but it doesn't have to maintain a separate representation of the original values "off to the side", it just tracks those as fields in the custom proxy class.

So, with all that said, because we have no change tracking right now it's the call to `Set` that flags to the state manager that the piece of state you're setting has changed (i.e. a dirty bit) and should be persisted back to the store. If people find having to call `Set` an annoyance then I believe that implementing basic snapshot change tracking is a medium effort and would provide a good solution to the problem, tho without solving the dirty flag problem, we'd pay the price of checking all outstanding state POCOs for diffs, but that's something I think we could live with on initial release.

Finally, for completeness, the way this is solved on the JS side is that the state management internals serialize the state objects to a JSON string, hash that JSON string and compare the hashes. This seems extremely costly to me. On the .NET side, where I understand internals best, this would create a massive amount of garbage which, memory usage aside, is kind of the worst thing you can do for performance... the garbage collector is not your friend. 😕

> The default state store is a nice idea, but I wonder if this will lead to people forgetting to add the default store and then not explicitly referencing one when adding middleware. I kind of favor the traditional approach of specifying your state store for each piece of state middleware if I am honest. At least this way the user cannot fall into the trap of not specifying one and means its also really clear which store is being used for each piece of middleware.

If you "forget" and then try to request a state manager that wasn't explicitly bound to a named state store you will very quickly get an `InvalidOperationException` with a message that explains that no explicit store was set and no default store was set either when you try to access that state. 

Truth be told, if people feel the runtime exception might be a little late, we could consider adding a validation API that checks that all configured state managers are mapped to available stores at middleware configuration time. Something like:

```CSharp
middleware.Use(
    new StateManagementMiddleware()
         // ... all config here ...
         .Validate());
```
Where the call to `Validate` walks through all the mappings and makes sure they line up, otherwise it throws an exception of some kind right then and there.

**Comment by garypretty** [comment]
If I had to choose been the InvalidOperationException when you tried to retrieve state and the validate method, I would favor the former as I don't like the addition of something else for the developer to call / add.

**Comment by drub0y** [comment]
I agree... if it was a choice that's the one I'd choose for sure, but, to be clear, I wouldn't remove that runtime check even if I were to add `Validate` to the middleware, it would just enable people who did use that feature to find out quicker.

**Comment by tomlm** [comment]
Wow, you've been a busy beaver. 

A couple of comments:
 I think I like the general approach, but as I said in the CR I think .Set() is problematic because
the lifetime semantics of who is changing an object is not clear, and so you are pushing on to the developer to make sure that .Set() doesn't clobber over the top of someone elses change. (I think)

I understand the concern about perf of change tracking, but I think the only real thing that needs to be tracked is that the object has changed. Obviously we could continue to do the the thing we do now which is to compare serialized forms (on the assumption serialization is highly optimized and cross platform expression of reflecting all properties), and I would say that's what we should do unless we can come up with something more clever.

I wonder if instead of .Set() we simply add a .SetChanged() or .MarkDirty() or whatever way to signify object has been changed without passing in the object.  So .Get<MyState>() would have companion .SetChanged<MyState>().  This would mean we only need to track dirty objects and not dirty properties which seems doable to me, and we wouldn't be setting a semantic where I am passing in a possible newed up object that would clobber someone elses copy. 

That said, while I like how this addresses many of the turds in our current state implementation (IStoreItem, FlexObject, non-parallel fetches) etc. I am concerned about how late a change this is.  I tip my hat to you that you have implemented in C#...how long do you think it would take to do the same in node?

**Comment by drub0y** [comment]
> I think I like the general approach, but as I said in the CR I think .Set() is problematic because
the lifetime semantics of who is changing an object is not clear, and so you are pushing on to the developer to make sure that .Set() doesn't clobber over the top of someone elses change. (I think)

Yes, that is correct. I did reply to your comment on the PR about this. Give that a read and tell me if that makes it any clearer. In short, in the PR comment I mention each state namespace is indeed a giant dictionary that all participants need to play nicely within. However, there are also many options provided to help people not shoot themselves in the foot including a full escape hatch of creating your own custom namespace.

> Obviously we could continue to do the the thing we do now which is to compare serialized forms (on the assumption serialization is highly optimized and cross platform expression of reflecting all properties), and I would say that's what we should do unless we can come up with something more clever.

So, just to be clear the JS side does the serialization + hash comparison, but .NET didn't do anything like this yet. The `BotState<StateT>` class indiscriminately takes the `StateT`, shoves it into a fresh `StoreItems` under the calculated key and hands it to the `IStore` which does no kind of analysis and all the implementations effectively send it to the underlying storage mechanism only doing `ETag` concurrency checks when applicable.

I'd have to stew on this for a bit because, right now you don't even have an original instance of the object you can compare against because the state manager is only working with a "materialized" `TState` instance. The original values are kept in the `StateStorageEntry` in whatever proprietary format the underlying storage provider chose to keep them in. So, we'd have to expose some kind of abstraction for the provider to do dirty checking instead. 

Honestly, most of the impact to make this happen is internal to the state manager <-> state provider layer and then we just remove the `Set` API and devs would just delete those calls from their code because they wouldn't need to do anything any more. You can even take an approach where you just make `Set` no-op under the covers and mark it as obsolete and remove it from the API down the road. This bolsters the case for adding an explicit `Add` method now though... read on.

> I wonder if instead of .Set() we simply add a .SetChanged() or .MarkDirty() or whatever way to signify object has been changed without passing in the object. So .Get() would have companion .SetChanged(). This would mean we only need to track dirty objects and not dirty properties which seems doable to me, and we wouldn't be setting a semantic where I am passing in a possible newed up object that would clobber someone elses copy.

Yeah so that's exactly what `Set` does right now: flags a dirty bit. Then `StateManager::SaveChanges` only submits those with the `IsDirty == true` to the store for persistence. If we change the name, which I'm not adverse to if that's the consensus, the only thing we'd need to do is make sure to add an explicit `Add` method for when you want to add new state as `Set` was currently handling both add and updates. As mentioned above, if we're already planning to obsolete the `Set` method, it only makes sense to add the `Add` method ASAP and guide people towards using that day one for new state values.

**Comment by garypretty** [comment]
@tomlm I get that this is a late change and the concern around this, but I honestly think this is just one of those things that needs to be addressed. In my opinion, whilst state management works right now to a degree (I closed a PR yesterday to address an issue on the c# side) I think that some of the areas this PR addresses will be essential for many use cases and maybe we just haven't seen the critical mass of the development on the Sdk to play that out.

I'm sure you probably agree with the above, but just wanted to say explicitly as I feel really strongly that this area should be addressed now.

**Comment by garypretty** [comment]
@billba Would be great to get your input here as I know we have been discussing the above elsewhere (as well as the small PR, which I have now closed, that would have brought the .net SDK inline with the JS for use of bot state directly).  
@Stevenic cc'ing you as we are now discussing the impact on implementing the above on the JS side..

**Comment by tomlm** [comment]
Argh.  I didn't realize the C# code was not doing the change detection...I know that the node stuff does and assumed that got ported the same. ☹️ 

Yeah, the biggest reason to take this now is that while it is relatively easy to change the names of things  (because a developer just has to refactor their code a bit) changing storage requires them to throw away their data and/or have to deal with migration which is a huge pain to push on our community.

@Stevenic we really need to think about taking this.

**Comment by drub0y** [comment]
Sooo, I started to deal with the rebasing and the current design of the `TranslationMiddleware` and `LocaleConverterMiddleware` does not play well with this because they each currently take a `Func<ITurnContext, something>` which is expected to return them some value, but the problem is all of the sample/test code is loading that `something` from state which is now an `async` API call. I can easily go ahead and change these signatures to `Func<ITurnContext, Task<something>>` , but I wanted to make sure I let you guys know in advance.

Now, that would just get me to the point where I can have a clean build and everything continues to work pretty much the way it is, but then we probably should have a discussion about the design of these altogether. This does not "feel" right in that it is not playing within the bounds of the framework correctly. IMHO, these middleware implementations should be managing their own state (now that they can actually do that!) and exposing their values to downstream logic via services + extension methods.

**Comment by drub0y** [comment]
I have made the changes mentioned in the *Step 1* section in the reply above to the `TranslationMiddleware` and `LocaleConverterMiddleware` and pushed the rebased code up. All tests passing locally... will see what happens with build caused by the push completes.

I don't have a way to test the sample projects because I don't have a LUIS model for them. If someone from the team wants to pull the latest from here and run them to make sure they still work that would be much appreciated.

**Comment by garypretty** [comment]
@drub0y I have pulled down your latest and tested the LUIS sample against my own model and it is working :) 👍

**Comment by billba** [comment]
@garypretty I have consulted closely with @drub0y on this proposal. It has my blessing.

**Comment by garypretty** [comment]
@billba great. Thought so. Just closing the loop :)

**Comment by drub0y** [comment]
So, any thoughts about the open API design issue I called out regarding moving towards a single extension method, `.State()` that returns a service with the various types of state from there as opposed to multiple extension methods? 'Cause I'm kind of leaning in favor of that nowadays and if there's enough consensus to head towards that approach I could easily make that change. I just don't want to do it if someone sees the existing, multiple extension methods approach as somehow being better.

**Comment by drub0y** [comment]
I updated the original issue text, but just calling out that the open discussion around optimistic concurrency has been resolved at this time:

> * Should optimistic concurrency be _required_? Right now the concept of `ETag` is present on the `IStateStorageEntry` interface and implies somehow that all stores _might_ provide optimistic concurrency. This is by no means easy (e.g. file system), nor is it always necessary for people, sometimes Last Write Wins is just fine. 

After some internal debate, we are removing support for optimistic concurrency (i.e. `ETag`) from the state management APIs. It was deemed unnecessary due to the existing partitioning that naturally takes place in a conversation which should avoid most conflicts and because there's no good solution for what to do with the fact that an optimistic concurrency violation occurs. Instead favoring what would normally happen in a conversation which would be opportunities for confirmation and clarification with the user who would then have a chance to notice something might be out of alignment and correct it.

**Comment by vishwacsena** [comment]
State management is in and will be available as part of SDK v4 stable release.
